File: code/experiments/experiment.cpp
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/twist.hpp>
#include <sensor_msgs/msg/image.hpp>

#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>

#include <chrono>
#include <cmath>
#include <string>
#include <vector>

#include <curl/curl.h>
#include <nlohmann/json.hpp>

using namespace std::chrono_literals;
using json = nlohmann::json;

// ----------------------------
// CONFIG
// ----------------------------

static constexpr double TARGET_DISTANCE_M = 1.83;
static constexpr double FORWARD_SPEED = 0.25;
static constexpr double TURN_SPEED = 0.4;
static constexpr double STOP_DISTANCE_M = 1.2;
static constexpr double DEPTH_FOV_RATIO = 0.3;

static const std::string VLM_ENDPOINT = "http://localhost:5000/vlm";

// ----------------------------
// CURL HELPERS
// ----------------------------

static size_t WriteCallback(void* contents, size_t size, size_t nmemb, void* userp)
{
    ((std::string*)userp)->append((char*)contents, size * nmemb);
    return size * nmemb;
}

// ----------------------------
// NODE
// ----------------------------

class Go2Navigator : public rclcpp::Node
{
public:
    Go2Navigator() : Node("go2_yolo_vlm_nav")
    {
        rgb_sub_ = create_subscription<sensor_msgs::msg::Image>(
            "/camera/color/image_raw", 10,
            std::bind(&Go2Navigator::rgbCallback, this, std::placeholders::_1));

        depth_sub_ = create_subscription<sensor_msgs::msg::Image>(
            "/camera/depth/image_raw", 10,
            std::bind(&Go2Navigator::depthCallback, this, std::placeholders::_1));

        cmd_pub_ = create_publisher<geometry_msgs::msg::Twist>("/cmd_vel", 10);

        start_time_ = now();

        RCLCPP_INFO(get_logger(), "Go2 C++ YOLO + VLM Navigator started");
    }

    void step()
    {
        if (rgb_frame_.empty() || depth_frame_.empty())
            return;

        double elapsed = (now() - start_time_).seconds();
        double distance_traveled = elapsed * FORWARD_SPEED;

        if (distance_traveled >= TARGET_DISTANCE_M)
        {
            stop();
            RCLCPP_INFO(get_logger(), "Reached target distance");
            rclcpp::shutdown();
            return;
        }

        double depth_ahead = getForwardDepth();

        std::string direction = "forward";

        if (depth_ahead < STOP_DISTANCE_M)
        {
            // --------------------------------
            // YOLO PLACEHOLDER (C++ inference)
            // --------------------------------
            bool yolo_has_box = false;
            double box_center_x = 0.0;

            // TODO: Replace with OpenCV DNN / TensorRT YOLO
            // If YOLO detects something, set yolo_has_box = true

            if (yolo_has_box)
            {
                direction = (box_center_x < rgb_frame_.cols / 2) ? "right" : "left";
            }
            else
            {
                direction = queryVLM(rgb_frame_);
            }
        }

        executeMotion(direction);
    }

private:
    // ----------------------------
    // ROS
    // ----------------------------

    rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr rgb_sub_;
    rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr depth_sub_;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr cmd_pub_;

    rclcpp::Time start_time_;

    cv::Mat rgb_frame_;
    cv::Mat depth_frame_;

    // ----------------------------
    // CALLBACKS
    // ----------------------------

    void rgbCallback(const sensor_msgs::msg::Image::SharedPtr msg)
    {
        rgb_frame_ = cv_bridge::toCvCopy(msg, "bgr8")->image;
    }

    void depthCallback(const sensor_msgs::msg::Image::SharedPtr msg)
    {
        cv::Mat depth_mm = cv_bridge::toCvCopy(msg)->image;
        depth_mm.convertTo(depth_frame_, CV_32F, 1.0 / 1000.0);  // mm → meters
    }

    // ----------------------------
    // DEPTH
    // ----------------------------

    double getForwardDepth()
    {
        int h = depth_frame_.rows;
        int w = depth_frame_.cols;

        int cx1 = static_cast<int>(w * (0.5 - DEPTH_FOV_RATIO / 2));
        int cx2 = static_cast<int>(w * (0.5 + DEPTH_FOV_RATIO / 2));
        int cy1 = static_cast<int>(h * 0.4);
        int cy2 = static_cast<int>(h * 0.8);

        cv::Mat roi = depth_frame_(cv::Range(cy1, cy2), cv::Range(cx1, cx2));
        cv::Mat mask = roi > 0;

        std::vector<float> values;
        roi.copyTo(values, mask);

        if (values.empty())
            return std::numeric_limits<double>::infinity();

        std::nth_element(values.begin(),
                         values.begin() + values.size() / 10,
                         values.end());

        return values[values.size() / 10];
    }

    // ----------------------------
    // VLM QUERY
    // ----------------------------

    std::string queryVLM(const cv::Mat& image)
    {
        std::vector<uchar> buffer;
        cv::imencode(".jpg", image, buffer);

        std::string img_b64 = base64Encode(buffer);

        json payload = {
            {"image", img_b64},
            {"prompt",
             "You are assisting a blind user using a robot dog. "
             "Respond with one word only: left, right, forward, or stop."}
        };

        CURL* curl = curl_easy_init();
        if (!curl)
            return "stop";

        std::string response;
        struct curl_slist* headers = nullptr;
        headers = curl_slist_append(headers, "Content-Type: application/json");

        curl_easy_setopt(curl, CURLOPT_URL, VLM_ENDPOINT.c_str());
        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, payload.dump().c_str());
        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
        curl_easy_setopt(curl, CURLOPT_TIMEOUT, 5);

        curl_easy_perform(curl);
        curl_easy_cleanup(curl);

        auto j = json::parse(response, nullptr, false);
        std::string text = j.value("response", "");

        if (text.find("left") != std::string::npos) return "left";
        if (text.find("right") != std::string::npos) return "right";
        if (text.find("forward") != std::string::npos) return "forward";

        return "stop";
    }

    // ----------------------------
    // MOTION
    // ----------------------------

    void executeMotion(const std::string& direction)
    {
        geometry_msgs::msg::Twist cmd;

        if (direction == "forward")
            cmd.linear.x = FORWARD_SPEED;
        else if (direction == "left")
            cmd.angular.z = TURN_SPEED;
        else if (direction == "right")
            cmd.angular.z = -TURN_SPEED;

        cmd_pub_->publish(cmd);
    }

    void stop()
    {
        cmd_pub_->publish(geometry_msgs::msg::Twist());
    }

    // ----------------------------
    // BASE64
    // ----------------------------

    static std::string base64Encode(const std::vector<uchar>& data);
};

File: code/experiments/experiment.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node

from geometry_msgs.msg import Twist
from sensor_msgs.msg import Image
from cv_bridge import CvBridge

import cv2
import numpy as np
import torch
import time
import base64
import requests

from ultralytics import YOLO

# ----------------------------
# CONFIG
# ----------------------------

TARGET_DISTANCE_M = 1.83       # ~6 feet
FORWARD_SPEED = 0.25           # m/s (safe walking pace)
TURN_SPEED = 0.4               # rad/s
SIDE_STEP_TIME = 1.2           # seconds
STOP_DISTANCE_M = 1.2          # obstacle threshold
YOLO_CONF = 0.01               # VERY LOW confidence
DEPTH_FOV_RATIO = 0.3          # center depth window
OLLAMA_ENDPOINT = "http://localhost:5000/vlm"
VLM_URL = "http://localhost:11434/api/generate"
VLM_MODEL = "llava-phi3"

# ----------------------------
# NODE
# ----------------------------

class Go2Navigator(Node):
    def __init__(self):
        super().__init__("go2_yolo_vlm_nav")

        self.bridge = CvBridge()

        # Subscribers
        self.rgb_sub = self.create_subscription(
            Image,
            "/camera/color/image_raw",
            self.rgb_callback,
            10
        )

        self.depth_sub = self.create_subscription(
            Image,
            "/camera/depth/image_raw",
            self.depth_callback,
            10
        )

        # Publisher
        self.cmd_pub = self.create_publisher(Twist, "/cmd_vel", 10)

        # Models
        self.yolo = YOLO("yolov8n.pt")
        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
        self.yolo.to(self.device)

        # State
        self.rgb_frame = None
        self.depth_frame = None
        self.start_time = None
        self.distance_traveled = 0.0
        self.last_cmd_time = time.time()

        self.get_logger().info("Go2 YOLO + VLM Navigator Initialized")

    # ----------------------------
    # CALLBACKS
    # ----------------------------

    def rgb_callback(self, msg):
        self.rgb_frame = self.bridge.imgmsg_to_cv2(msg, desired_encoding="bgr8")

    def depth_callback(self, msg):
        depth = self.bridge.imgmsg_to_cv2(msg, desired_encoding="passthrough")
        self.depth_frame = depth.astype(np.float32) / 1000.0  # mm → meters

    # ----------------------------
    # CORE LOGIC
    # ----------------------------

    def get_forward_depth(self):
        """Return minimum depth in center of frame"""
        if self.depth_frame is None:
            return np.inf

        h, w = self.depth_frame.shape
        cx1 = int(w * (0.5 - DEPTH_FOV_RATIO / 2))
        cx2 = int(w * (0.5 + DEPTH_FOV_RATIO / 2))
        cy1 = int(h * 0.4)
        cy2 = int(h * 0.8)

        roi = self.depth_frame[cy1:cy2, cx1:cx2]
        roi = roi[roi > 0]

        if roi.size == 0:
            return np.inf

        return np.percentile(roi, 10)  # robust minimum

    def query_vlm(self, image):
        """
        Send RGB frame to local Flask VLM server (port 5000)
        Returns: 'left', 'right', 'forward', or 'stop'
        """

        # Encode image
        _, buffer = cv2.imencode(".jpg", image)
        image_b64 = base64.b64encode(buffer).decode("utf-8")

        prompt = (
            "You are assisting a blind user using a robot dog. "
            "Based on the image, determine if there is an obstacle directly ahead. "
            "Respond with ONE word only: left, right, forward, or stop."
        )

        payload = {
            "image": image_b64,
            "prompt": prompt
        }

        try:
            r = requests.post(OLLAMA_ENDPOINT, json=payload, timeout=5)

            if not r.ok:
                return "stop"

            response = r.json().get("response", "").lower()

            # Hard safety filter
            if "left" in response:
                return "left"
            elif "right" in response:
                return "right"
            elif "forward" in response:
                return "forward"
            else:
                return "stop"

        except Exception as e:
            self.get_logger().warn(f"VLM error: {e}")
            return "stop"


    def step(self):
        if self.rgb_frame is None or self.depth_frame is None:
            return

        if self.start_time is None:
            self.start_time = time.time()

        elapsed = time.time() - self.start_time
        self.distance_traveled = elapsed * FORWARD_SPEED

        if self.distance_traveled >= TARGET_DISTANCE_M:
            self.stop()
            self.get_logger().info("Reached target distance")
            rclpy.shutdown()
            return

        depth_ahead = self.get_forward_depth()

        # YOLO inference
        results = self.yolo(self.rgb_frame, conf=YOLO_CONF, verbose=False)[0]

        obstacle_detected = False
        direction = "forward"

        if depth_ahead < STOP_DISTANCE_M:
            obstacle_detected = True

        if obstacle_detected:
            if results.boxes is not None and len(results.boxes) > 0:
                box = results.boxes.xyxy[0].cpu().numpy()
                cx = (box[0] + box[2]) / 2

                if cx < self.rgb_frame.shape[1] / 2:
                    direction = "right"
                else:
                    direction = "left"
            else:
                direction = self.query_vlm(self.rgb_frame)

        self.execute_motion(direction)

    # ----------------------------
    # MOTION COMMANDS
    # ----------------------------

    def execute_motion(self, direction):
        cmd = Twist()

        if direction == "forward":
            cmd.linear.x = FORWARD_SPEED

        elif direction == "left":
            cmd.angular.z = TURN_SPEED

        elif direction == "right":
            cmd.angular.z = -TURN_SPEED

        else:  # stop
            pass

        self.cmd_pub.publish(cmd)

    def stop(self):
        self.cmd_pub.publish(Twist())


# ----------------------------
# MAIN
# ----------------------------

def main():
    rclpy.init()
    node = Go2Navigator()

    rate = node.create_rate(10)

    try:
        while rclpy.ok():
            rclpy.spin_once(node)
            node.step()
            rate.sleep()
    except KeyboardInterrupt:
        pass
    finally:
        node.stop()
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()
